### PR TITLE
YD-526 Register app last opened date on device

### PR DESCRIPTION
**Pull request by Bert-R**
Along with this:
* Moved the path for the app last opened event from the UserController to the DeviceController
* Moved the implementation from the UserService to the DeviceService
* Extended the unit test
* Created an integration test
* Added extra assertions to the integration tests
* Corrected the assertEquals method for a string and a LocalDate in CommonAssertions

### CODE REVIEW

File: appservice/src/intTest/groovy/nu/yona/server/DeviceTest.groovy [hunk 1]
```diff
@@ -178,16 +182,94 @@ class DeviceTest extends AbstractAppServiceIntegrationTest
...
+		def response = appService.createResourceWithPassword(richard.postOpenAppEventUrl, "{}", richard.password)
+
+		then:
+		assertResponseStatusOk(response)
+
+		cleanup:
+		appService.deleteUser(richard)
+	}
+
+	def 'Richard posts app opened event with a valid app version'()
```
Comment (aristotelos) [inline appservice/src/intTest/groovy/nu/yona/server/DeviceTest.groovy]: "valid app version"-> "valid operating system and app version"

File: appservice/src/main/java/nu/yona/server/device/rest/DeviceController.java [hunk 2]
```diff
@@ -211,6 +227,20 @@ private void logLongAppActivityBatch(MessageType messageType, UUID userId, AppAc
...
 
+	public static Link getPostOpenAppEventLink(UUID userId, UUID deviceId)
+	{
+		try
+		{
+			ControllerLinkBuilder linkBuilder = linkTo(
+					methodOn(DeviceController.class).postOpenAppEvent(Optional.empty(), userId, deviceId, null));
+			return linkBuilder.withRel("postOpenAppEvent");
+		}
+		catch (SecurityException e)
```
Comment (aristotelos) [inline appservice/src/main/java/nu/yona/server/device/rest/DeviceController.java]: Why would this ever give a `SecurityException`?

File: core/src/main/java/nu/yona/server/device/service/DeviceService.java [hunk 3]
```diff
@@ -243,4 +255,94 @@ public void updateOperatingSystem(UUID userId, UUID deviceId, OperatingSystem op
...
+			case UNKNOWN:
+				// Everything is OK for as long as the app didn't confirm the operating system
+				return;
+			default:
+				throw new IllegalArgumentException("Unknown operating system: " + operatingSystem);
+		}
+		assertNotOlderThan(operatingSystem, minAppVersion, parseVersionString(appVersionStr));
+	}
+
+	private void assertNotOlderThan(OperatingSystem operatingSystem, String minAppVersionStr, Version appVersion)
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/device/service/DeviceService.java]: Please switch `operatingSystem` to the last parameter position.

File: core/src/main/resources/messages/messages_nl.properties [hunk 4]
```diff
@@ -113,6 +113,9 @@ error.device.cannot.delete.last.one=Kan laatste apparaat niet verwijderen. Appar
 error.device.collection.empty=De geanonymiseerde gebruiker met ID ''{0}'' heeft geen apparaten. Dit is geen geldige toestand.
 error.device.not.found.index=De geanonymiseerde gebruiker met ID ''{0}'' heeft geen apparaat de index ''{1}''
 error.device.name.already.exists=Er bestaat al een apparaat met de naam ''{0}''
+error.device.app.version.not.supported=Yona app is te oud en moet worden ge�pdatet. De huidige versie is {2} maar de oudste ondersteunde versie voor {0} is {1}
```
Comment (aristotelos) [inline core/src/main/resources/messages/messages_nl.properties]: geüpdatet -> geüpdate

File: appservice/src/intTest/groovy/nu/yona/server/DeviceTest.groovy (hunk 1 above)
Comment (Bert-R) [reply appservice/src/intTest/groovy/nu/yona/server/DeviceTest.groovy]: Done.

File: appservice/src/main/java/nu/yona/server/device/rest/DeviceController.java (hunk 2 above)
Comment (Bert-R) [reply appservice/src/main/java/nu/yona/server/device/rest/DeviceController.java]: No idea. It came this way from ``UserController``, but it isn't necessary. Removed it.

File: core/src/main/java/nu/yona/server/device/service/DeviceService.java (hunk 3 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/device/service/DeviceService.java]: Done

File: core/src/main/resources/messages/messages_nl.properties (hunk 4 above)
Comment (Bert-R) [reply core/src/main/resources/messages/messages_nl.properties]: I feel the same, but alas, the spelling I used is correct. See https://nl.wiktionary.org/wiki/updaten/vervoeging

**Comment by Bert-R** [comment]
Pushed an update

**Pull request merged**